ex-4-quicksort/ex-4-4-recursive-binary-search.js

```diff
@@ -0,0 +1,32 @@
+//Exercise 4.4
+//Recursive Binary Search
+
+function binarySearch( arr, val ) {
+    let low = 0;
+    let high = arr.length - 1;
+    let mid = Math.floor( ( low + high ) / 2 );
+
+    if ( arr[mid] === val ) {
+        return "Found - " + arr[mid];
+    }
+
+    if ( arr[0] === arr[high] ) {
+        return "Not found";
+    }
+
+    if ( arr[mid] > val ) {
+        return binarySearch( [...arr.slice(0, mid)], val ); // rerun search with lower half of array
+    } else {
+        return binarySearch( [...arr.slice(mid + 1)], val ); // rerun search with higher half of array
+    }
```

Comment:
Now that you've written `binarySearch` with and without recursion, which do you think is easier to explain? What about which was easier to write?

Reply:
I feel like it's both easier to explain and write with recursion.  Explaining with recursion would go something like: 

Check if the middle value matches the input value.  If the middle value is greater than what were looking for, run the function again with the only the values less than the middle value.  Otherwise, run the function again with all values higher than than the middle value.  

Without recursion, the function is running a loop where it's conditions are updated inside the loop.  It's confusing to read and to explain.